SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,278 @@
+
+	Trace_EventData* EventData = Trace_EventData_Create();
+
+	for (const auto& Pair : SpatialTraceEvent.Data)
+	{
+		auto KeySrc = StringCast<ANSICHAR>(*Pair.Key);
+		const ANSICHAR* Key = KeySrc.Get();
+		auto ValueSrc = StringCast<ANSICHAR>(*Pair.Value);
+		const ANSICHAR* Value = ValueSrc.Get();
+		Trace_EventData_AddStringFields(EventData, 1, &Key, &Value);
+	}
+
+	// Frame counter
+	{
+		const char* FrameCountStr = "FrameNum";
+		char TmpBuffer[64];
+		FCStringAnsi::Sprintf(TmpBuffer, "%" PRIu64, GFrameCounter);
+		const char* TmpBufferPtr = TmpBuffer;
+		Trace_EventData_AddStringFields(EventData, 1, &FrameCountStr, &TmpBufferPtr);
+	}
+
+	TraceEvent.data = EventData;
+	Trace_EventTracer_AddEvent(EventTracer, &TraceEvent);
+	Trace_EventData_Destroy(EventData);
+
+	return CurrentSpanId;
+}
+
+bool SpatialEventTracer::IsEnabled() const
+{
+	return bEnabled; // Trace_EventTracer_IsEnabled(EventTracer);
+}
+
+void SpatialEventTracer::Enable(const FString& FileName)
+{
+	Trace_EventTracer_Parameters parameters = {};
+	parameters.user_data = this;
+	parameters.callback = &SpatialEventTracer::TraceCallback;
+	EventTracer = Trace_EventTracer_Create(&parameters);
+	Trace_EventTracer_Enable(EventTracer);
+	bEnabled = true;
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	FString EventTracePath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
+	FString AbsLogPath;
+	if (FParse::Value(FCommandLine::Get(), TEXT("eventLogPath="), AbsLogPath, false))
+	{
+		EventTracePath = FPaths::GetPath(AbsLogPath);
+	}
+	const FString FolderPath = EventTracePath;
+
+	const FString FullFileName = FString::Printf(TEXT("EventTrace_%s_%s.trace"), *FileName, *FDateTime::Now().ToString());
+	const FString FilePath = FPaths::Combine(FolderPath, FullFileName);
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
+	{
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
+	}
+}
+
+void SpatialEventTracer::StreamDeleter::operator()(worker::c::Io_Stream* StreamToDestroy) const
+{
+	Io_Stream_Destroy(StreamToDestroy);
+}
+
+void SpatialEventTracer::Disable()
+{
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
+
+	Trace_EventTracer_Disable(EventTracer);
+	Trace_EventTracer_Destroy(EventTracer);
+
+	bEnabled = false;
+	Stream = nullptr;
+}
+
+void SpatialEventTracer::ComponentAdd(const Worker_Op& Op)
+{
+	SpanIdStore.ComponentAdd(Op);
+}
+
+void SpatialEventTracer::ComponentRemove(const Worker_Op& Op)
+{
+	SpanIdStore.ComponentRemove(Op);
+}
+
+void SpatialEventTracer::ComponentUpdate(const Worker_Op& Op)
+{
+	const Worker_ComponentUpdateOp& ComponentUpdateOp = Op.op.component_update;
+	EntityComponentId Id(ComponentUpdateOp.entity_id, ComponentUpdateOp.update.component_id);
```

Comment:
nit - style I was trying to go with throught spatial view (and so everywhere this type is used) is to use = {} for setting POD types. 
So in this case `EntityComponentId Id = {ComponentUpdateOp.entity_id, ComponentUpdateOp.update.component_id}`
I think adding a constructor to the type is fine if you want to use it in emplace calls (or any other forwarding construction) but I'd rather not have it for this.